File: lua/lockfile/parse/npm.lua
--- package-lock.json / npm-shrinkwrap.json -> normalized model.
---
--- Supports lockfileVersion 2/3 (the `packages` map keyed by install path) and
--- falls back to the legacy v1 `dependencies` tree. Dependency edges in npm are
--- semver *ranges*, not resolved versions, so edges carry only a name and are
--- resolved against the installed set by name during analysis.

local json = require("lockfile.parse.json")
local model = require("lockfile.model")

local M = {}

local NODE_MODULES = "node_modules/"

--- Derive a package name from a `packages` install path key.
--- "node_modules/a/node_modules/@s/b" -> "@s/b". Returns nil for the root "".
---@param path string
---@return string?
local function name_from_path(path)
  if path == "" then
    return nil
  end
  local idx
  local start = 1
  while true do
    local f = path:find(NODE_MODULES, start, true)
    if not f then
      break
    end
    idx = f
    start = f + 1
  end
  if idx then
    return path:sub(idx + #NODE_MODULES)
  end
  -- A path without node_modules is a local workspace package.
  return nil
end

--- Append dependency edges (by name) from an npm dependency-spec map.
---@param map any
---@param out lockfile.Dep[]
local function collect_edges(map, out)
  if type(map) ~= "table" then
    return
  end
  for name in pairs(map) do
    if type(name) == "string" then
      out[#out + 1] = { name = name }
    end
  end
end

--- Build from a lockfileVersion 2/3 `packages` map.
---@param lf lockfile.Lockfile
---@param packages table
local function build_v3(lf, packages)
  local root_names = {}

  for path, entry in pairs(packages) do
    if type(path) == "string" and type(entry) == "table" then
      if path == "" then
        -- The project root: its declared deps are the direct dependencies.
        collect_edges(entry.dependencies, root_names)
        collect_edges(entry.devDependencies, root_names)
        collect_edges(entry.optionalDependencies, root_names)
      else
        local name = name_from_path(path)
        if name == nil and type(entry.name) == "string" then
          -- workspace package addressed by directory path
          name = entry.name
        end
        if name and type(entry.version) == "string" then
          local deps = {}
          collect_edges(entry.dependencies, deps)
          collect_edges(entry.optionalDependencies, deps)
          local pkg = lf:add({
            name = name,
            version = entry.version,
            source = entry.resolved,
            checksum = entry.integrity,
            deps = deps,
            dev = entry.dev == true,
            optional = entry.optional == true,
          })
          -- workspace packages (path without node_modules) are roots
          if name_from_path(path) == nil then
            lf:add_root(pkg.id)
          end
        end
      end
    end
  end

  -- Resolve declared root names to concrete installed package ids.
  for _, dep in ipairs(root_names) do
    local list = lf.by_name[dep.name]
    if list and list[1] then
      lf:add_root(list[1].id)
    end
  end
end

--- Recursively build from a legacy v1 `dependencies` tree.
---@param lf lockfile.Lockfile
---@param deps table
---@param roots boolean   # whether this level's entries are project roots
local function build_v1(lf, deps, roots)
  for name, entry in pairs(deps) do
    if type(name) == "string" and type(entry) == "table" and type(entry.version) == "string" then
      local edges = {}
      collect_edges(entry.requires, edges)
      local pkg = lf:add({
        name = name,
        version = entry.version,
        source = entry.resolved,
        checksum = entry.integrity,
        deps = edges,
        dev = entry.dev == true,
        optional = entry.optional == true,
      })
      if roots then
        lf:add_root(pkg.id)
      end
      if type(entry.dependencies) == "table" then
        build_v1(lf, entry.dependencies, false)
      end
    end
  end
end

--- Build a normalized lockfile from package-lock.json source text.
---@param src string
---@return lockfile.Lockfile
function M.build(src)
  local data = json.parse(src)
  local lf = model.new("npm")
  if type(data) ~= "table" then
    return lf
  end
  if data.lockfileVersion ~= nil then
    lf.format_version = tostring(data.lockfileVersion)
  end

  if type(data.packages) == "table" then
    build_v3(lf, data.packages)
  elseif type(data.dependencies) == "table" then
    build_v1(lf, data.dependencies, true)
  end

  return lf
end

return M

File: src/parse/npm.rs
//! package-lock.json / npm-shrinkwrap.json -> normalized model.
//!
//! Supports lockfileVersion 2/3 (the `packages` map keyed by install path) and
//! the legacy v1 `dependencies` tree. npm dependency edges are semver ranges,
//! not resolved versions, so edges carry only a name.

use std::collections::HashMap;

use serde::Deserialize;
use serde_json::Value;

use crate::model::{Dep, Lockfile, Package};

const NODE_MODULES: &str = "node_modules/";

#[derive(Deserialize)]
struct NpmLock {
    #[serde(rename = "lockfileVersion")]
    lockfile_version: Option<Value>,
    packages: Option<HashMap<String, NpmPkg>>,
    dependencies: Option<HashMap<String, NpmV1Dep>>,
}

#[derive(Deserialize)]
struct NpmPkg {
    name: Option<String>,
    version: Option<String>,
    resolved: Option<String>,
    integrity: Option<String>,
    #[serde(default)]
    dependencies: HashMap<String, Value>,
    #[serde(rename = "devDependencies", default)]
    dev_dependencies: HashMap<String, Value>,
    #[serde(rename = "optionalDependencies", default)]
    optional_dependencies: HashMap<String, Value>,
    #[serde(default)]
    dev: bool,
    #[serde(default)]
    optional: bool,
    #[serde(default)]
    link: bool,
}

#[derive(Deserialize)]
struct NpmV1Dep {
    version: String,
    resolved: Option<String>,
    integrity: Option<String>,
    #[serde(default)]
    requires: HashMap<String, Value>,
    #[serde(default)]
    dependencies: HashMap<String, NpmV1Dep>,
    #[serde(default)]
    dev: bool,
    #[serde(default)]
    optional: bool,
}

/// Derive a package name from a `packages` install-path key.
/// "node_modules/a/node_modules/@s/b" -> Some("@s/b"); "" -> None.
fn name_from_path(path: &str) -> Option<&str> {
    if path.is_empty() {
        return None;
    }
    path.rfind(NODE_MODULES)
        .map(|idx| &path[idx + NODE_MODULES.len()..])
}

/// Dependency edges (by name) from an npm dependency-spec map.
fn edges_from(map: &HashMap<String, Value>) -> Vec<Dep> {
    map.keys().map(Dep::by_name).collect()
}

fn version_string(v: &Value) -> String {
    match v {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    }
}

fn build_v3(lf: &mut Lockfile, packages: HashMap<String, NpmPkg>) {
    // Direct dependency names declared by the project root entry.
    let mut root_names: Vec<String> = Vec::new();
    // name -> id for resolving root names to concrete installs.
    let mut by_name: HashMap<String, String> = HashMap::new();

    for (path, entry) in &packages {
        if path.is_empty() {
            root_names.extend(entry.dependencies.keys().cloned());
            root_names.extend(entry.dev_dependencies.keys().cloned());
            root_names.extend(entry.optional_dependencies.keys().cloned());
            continue;
        }

        let name = name_from_path(path)
            .map(str::to_string)
            .or_else(|| entry.name.clone());
        let (Some(name), Some(version)) = (name, entry.version.clone()) else {
            continue;
        };

        let mut pkg = Package::new(&name, &version);
        let mut deps = edges_from(&entry.dependencies);
        deps.extend(edges_from(&entry.optional_dependencies));
        pkg.deps = deps;
        pkg.source = entry.resolved.clone();
        pkg.checksum = entry.integrity.clone();
        pkg.dev = entry.dev;
        pkg.optional = entry.optional;

        by_name
            .entry(name.clone())
            .or_insert_with(|| pkg.id.clone());

        // Workspace packages (path without node_modules) are roots.
        if name_from_path(path).is_none() && !entry.link {
            lf.add_root(pkg.id.clone());
        }
        lf.push(pkg);
    }

    for name in root_names {
        if let Some(id) = by_name.get(&name) {
            lf.add_root(id.clone());
        }
    }
}

fn build_v1(lf: &mut Lockfile, deps: HashMap<String, NpmV1Dep>, roots: bool) {
    for (name, entry) in deps {
        let mut pkg = Package::new(&name, &entry.version);
        pkg.deps = edges_from(&entry.requires);
        pkg.source = entry.resolved;
        pkg.checksum = entry.integrity;
        pkg.dev = entry.dev;
        pkg.optional = entry.optional;
        if roots {
            lf.add_root(pkg.id.clone());
        }
        lf.push(pkg);
        build_v1(lf, entry.dependencies, false);
    }
}

pub fn parse(src: &str) -> Result<Lockfile, String> {
    let doc: NpmLock =
        serde_json::from_str(src).map_err(|e| format!("invalid package-lock.json: {e}"))?;

    let mut lf = Lockfile::new("npm");
    lf.format_version = doc.lockfile_version.as_ref().map(version_string);

    if let Some(packages) = doc.packages {
        build_v3(&mut lf, packages);
    } else if let Some(deps) = doc.dependencies {
        build_v1(&mut lf, deps, true);
    }

    Ok(lf)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_v3_packages_and_roots() {
        let src = r#"
        {
          "name": "myapp",
          "lockfileVersion": 3,
          "packages": {
            "": { "name": "myapp", "dependencies": { "express": "^4.18.0" } },
            "node_modules/express": {
              "version": "4.18.2",
              "resolved": "https://registry.npmjs.org/express/-/express-4.18.2.tgz",
              "integrity": "sha512-abc",
              "dependencies": { "accepts": "~1.3.8" }
            },
            "node_modules/accepts": { "version": "1.3.8", "integrity": "sha512-def" }
          }
        }
        "#;
        let lf = parse(src).unwrap();
        assert_eq!(lf.format_version.as_deref(), Some("3"));
        assert_eq!(lf.packages.len(), 2);
        assert_eq!(lf.roots, vec!["express@4.18.2"]);
        let express = lf.packages.iter().find(|p| p.name == "express").unwrap();
        assert_eq!(express.checksum.as_deref(), Some("sha512-abc"));
        assert_eq!(express.deps.len(), 1);
        assert_eq!(express.deps[0].name, "accepts");
    }

    #[test]
    fn scoped_name_from_nested_path() {
        assert_eq!(
            name_from_path("node_modules/a/node_modules/@s/b"),
            Some("@s/b")
        );
        assert_eq!(name_from_path(""), None);
        assert_eq!(name_from_path("packages/app"), None);
    }

    #[test]
    fn parses_v1_tree() {
        let src = r#"
        {
          "lockfileVersion": 1,
          "dependencies": {
            "express": {
              "version": "4.18.2",
              "integrity": "sha512-abc",
              "requires": { "accepts": "~1.3.8" },
              "dependencies": {
                "accepts": { "version": "1.3.8", "integrity": "sha512-def" }
              }
            }
          }
        }
        "#;
        let lf = parse(src).unwrap();
        assert_eq!(lf.packages.len(), 2);
        assert_eq!(lf.roots, vec!["express@4.18.2"]);
    }
}
